### PR TITLE
chore: update to typescript 4.5-rc

### LINKED ISSUE #3950: TypeScript 4.5 Support
https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-rc/

This issue is just to track all of the new features and their implementation state in this project.
As with all releases, we will not necessarily to support all features until closer to the full release when everything the features are stabilised.

*Please be patient.*

# `type` Modifiers on Import and Export Names

```ts
import { someFunc, type BaseType } from "./some-module.js";
//                 ^^^^^^^^^^^^^

type T = 1;
export { type T };
```

This will require AST changes.

Babel: https://github.com/babel/babel/pull/13802

- [x] PR: #4076

This will require updates to scope analysis:

- [x] PR: #4234

This will require updates to rules:

- `consistent-type-exports`
    - [x] PR: #4236
- `consistent-type-imports`
    - [x] PR: #4237
- `no-shadow`
    - [x] PR: #4239

# Private Field Presence Checks

```ts
class Person {
    #name: string;
    constructor(name: string) {
        this.#name = name;
    }

    equals(other: unknown) {
        return other &&
            typeof other === "object" &&
            #name in other &&
//          ^^^^^^^^^^^^^^
            this.#name === other.#name;
    }
}
```

This will require AST changes.

ESTree: https://github.com/estree/estree/blob/master/es2022.md#binaryexpression

- [x] PR: #4075

# Import Assertions

```ts
import obj from "./something.json" assert { type: "json" };
```

This will require AST changes.

ESTree: https://github.com/estree/estree/blob/master/stage3/import-assertions.md

- [x] PR: #4074, #4178

# `lib.d.ts` Changes

Will require regeneration of `@typescript-eslint/scope-manager` types.

- [x] PR: #4240

# Better Editor Support for Unresolved Types

This might break a few tests due to the type printer printing something that isn't `any`.

----

# Other changes with no impact to us

- Supporting lib from node_modules
- The `Awaited` Type and `Promise` Improvements
- Template String Types as Discriminants
- `--module es2022`
- Tail-Recursion Elimination on Conditional Types
- Disabling Import Elision
- Faster Load Time with `realPathSync.native`
- New Snippet Completions
- Better Editor Support for Unresolved Types
- Inference Changes from `Awaited`
- Compiler Options Checking at the Root of `tsconfig.json`
- Restrictions on Assignability to Conditional Types

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## PR Checklist

-   [x] Addresses an existing issue: #3950 
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md) were taken

## Overview

updates to TypeScript 4.5 RC.

~~updates to TypeScript 4.5 Beta.~~

~~I know typescript-eslint has the policy that is to implement new features when the RC version is released. The purpose of this pull request is to quickly add support for new features when the RC version is released. So this is still draft.~~

### CODE REVIEW

**Comment by sosukesuzuki** [comment]
One test of typescript-estree fails maybe for https://github.com/microsoft/TypeScript/issues/46577

File: packages/typescript-estree/src/parser.ts [hunk 1]
```diff
@@ -30,12 +30,12 @@ const log = debug('typescript-eslint:typescript-estree:parser');
...
  * typescript-eslint monorepo
  */
-const SUPPORTED_TYPESCRIPT_VERSIONS = '>=3.3.1 <4.5.0';
+const SUPPORTED_TYPESCRIPT_VERSIONS = '>=3.3.1 <4.6.0';
 /*
  * The semver package will ignore prerelease ranges, and we don't want to explicitly document every one
  * List them all separately here, so we can automatically create the full string
  */
-const SUPPORTED_PRERELEASE_RANGES: string[] = ['4.4.0-beta', '4.4.1-rc'];
+const SUPPORTED_PRERELEASE_RANGES: string[] = ['4.5.0-beta'];
```
Comment (bradzacher) [inline packages/typescript-estree/src/parser.ts:38]: 4.5rc is out so we should add this

```suggestion
const SUPPORTED_PRERELEASE_RANGES: string[] = ['4.5.0-beta', '4.5.1-rc'];
```

**Comment by sosukesuzuki** [comment]
I've updated tests to avoid microsoft/TypeScript#46577 (a3a89a7)

**Review by bradzacher — APPROVED** [review]
thanks!

**Pull request merged**